cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,139 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @throws IllegalStateException if something unexpected happened during upgrade.
+   * @throws IOException if there was an IO error during initializing application class from artifact.
+   * @throws JsonIOException if there was an error in serializing or deserializing app config.
+   * @throws Exception if there was an exception during the upgrade of application. This exception will often wrap
+   *                   the actual exception
+   */
+  public ApplicationUpdateDetail upgradeApplication(ApplicationId appId, ProgramTerminator programTerminator)
+      throws Exception {
+    // Check if the current user has admin privileges on it before updating.
+    authorizationEnforcer.enforce(appId, authenticationContext.getPrincipal(), Action.ADMIN);
+    // check that app exists
+    ApplicationSpecification currentSpec = store.getApplication(appId);
+    if (currentSpec == null) {
+      LOG.info("Application {} not found for upgrade.", appId);
+      return new ApplicationUpdateDetail(appId, new NotFoundException(appId));
+    }
+    ArtifactId currentArtifact = currentSpec.getArtifactId();
+    NamespaceId currentArtifactNamespace =
+      ArtifactScope.SYSTEM.equals(currentArtifact.getScope()) ? NamespaceId.SYSTEM : appId.getParent();
+
+    // Get the artifact with latest version for upgrade.
+    List<ArtifactSummary> availableArtifacts =
+      artifactRepository.getArtifactSummaries(currentArtifactNamespace, currentArtifact.getName(), 1,
+                                              ArtifactSortOrder.DESC);
+    if (availableArtifacts.isEmpty()) {
+      // This should not be possible as at least the current used artifact should be there.
+      String error = String.format("No artifacts found for artifact id %s in namespace %s.", currentArtifact.getName(),
+                                   currentArtifactNamespace);
+      return new ApplicationUpdateDetail(appId, "Upgrade failed.", error);
+    }
+    // The latest version should be first (and only) value in the result.
+    ArtifactSummary candidateArtifact = availableArtifacts.get(0);
+    ArtifactVersion candidateArtifactVersion = new ArtifactVersion(candidateArtifact.getVersion());
+
+    // Current artifact should not have higher version than candidate artifact.
+    if (currentArtifact.getVersion().compareTo(candidateArtifactVersion) > 0) {
+      String error = String.format(
+        "Requested artifact version %s is older than current artifact version %s.", candidateArtifact.getVersion(),
```

Comment:
This is a confusing error message because the user didn't request any artifact version, all they did was make an upgrade call. Would be better to say something like:

The current artifact has a version higher than any existing artifact.